modules/unsupported/vector-mosaic/src/main/java/org/geotools/vectormosaic/VectorMosaicStore.java

```diff
@@ -0,0 +1,484 @@
+     * @param granule the granule to find the data store for
+     */
+    @SuppressWarnings("unchecked")
+    private void findDataStore(String typeName, VectorMosaicGranule granule) {
+        try {
+            if (granule.getConnProperties() != null) {
+                String granuleCacheKey = typeName + "_" + granule.getGranuleTypeName();
+                if (!GRANULE_STORE_CACHE.containsKey(granuleCacheKey)
+                        || isConnPropertiesChanged(granuleCacheKey, granule)) {
+                    Map params = propertiesToMap(granule.getConnProperties());
+                    if (preferredSPI != null) {
+                        DataStoreFactorySpi dataStoreFactorySpi = getSPI(preferredSPI);
+                        granule.setDataStore(dataStoreFactorySpi.createDataStore(params));
+                    } else {
+                        granule.setDataStore(DataStoreFinder.getDataStore(params));
+                    }
+                    GRANULE_STORE_CACHE.put(granuleCacheKey, granule);
+                } else {
+                    granule.setDataStore(GRANULE_STORE_CACHE.get(granuleCacheKey).getDataStore());
+                }
+
+            } else {
+                LOGGER.log(
+                        Level.WARNING,
+                        "Connection properties not found for Vector Mosaic granule {0}",
+                        granule.getName());
+            }
+        } catch (IOException e) {
+            LOGGER.log(Level.WARNING, "Could not find data store", e);
+        }
+    }
+
+    /**
+     * Checks if the connection properties have changed for the given granule
+     *
+     * @param granuleCacheKey the key to use to find the granule in the cache
+     * @param granule the granule to check the connection properties against
+     * @return true if the connection properties have changed, false otherwise
+     */
+    private boolean isConnPropertiesChanged(String granuleCacheKey, VectorMosaicGranule granule) {
+        if (granule.getConnProperties() == null) {
+            return true;
+        }
+        return !GRANULE_STORE_CACHE
+                .get(granuleCacheKey)
+                .getConnProperties()
+                .equals(granule.getConnProperties());
+    }
+
+    /**
+     * Returns the data store factory SPI for the given name
+     *
+     * @param preferredSPI the name of the SPI to find
+     * @return the SPI
+     */
+    private DataStoreFactorySpi getSPI(String preferredSPI) {
+        DataStoreFactorySpi dataStoreFactorySpi = null;
+        for (Iterator<DataStoreFactorySpi> it = DataStoreFinder.getAvailableDataStores();
+                it.hasNext(); ) {
+            DataStoreFactorySpi spi = it.next();
+            if (spi.getClass().getTypeName().equals(preferredSPI)) {
+                dataStoreFactorySpi = spi;
+                break;
+            }
+        }
+        return dataStoreFactorySpi;
+    }
+
+    public boolean isTolerant() {
```

Comment:
Delaying the calculation of feature types should also remove the need for this property.

Reply:
Fixed